File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        minima_idx = []
        prev = head
        curr = head.next
        nxt = curr.next
        idx = 2
        while(nxt):
            if(prev.val > curr.val < nxt.val):
                minima_idx.append(idx)
            
            elif(prev.val < curr.val > nxt.val):
                minima_idx.append(idx)
            
            idx += 1
            prev, curr, nxt = curr, nxt, nxt.next
        
        if(len(minima_idx) < 2):
            return [-1, -1]
        
        min_val = inf
        for i in range(1, len(minima_idx)):
            min_val = min(min_val, minima_idx[i] - minima_idx[i - 1])
        
        return [min_val, minima_idx[-1] - minima_idx[0]]

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func nodesBetweenCriticalPoints(head *ListNode) []int {
    var prev, curr = head, head.Next
    var idxs = []int{}
    var index = 1
    for curr.Next != nil {
        if prev.Val < curr.Val && curr.Val > curr.Next.Val {
            idxs = append(idxs, index)
        }
        
        if prev.Val > curr.Val && curr.Val < curr.Next.Val {
            idxs = append(idxs, index)
        }
        index++
        prev = curr
        curr = curr.Next
    }
    
    var res = []int{-1, -1}
    if len(idxs) < 2 { return res }
    
    res[0] = math.MaxInt
    res[1] = idxs[len(idxs) - 1] - idxs[0]
    for i := range len(idxs) - 1 {
        res[0] = min(res[0], idxs[i + 1] - idxs[i])
    }
    return res
}